File: pizza.rb
# Over engineered Ruby version, author: Rune Hammersland
class Pizza
  attr_reader :name, :weight, :price

  def initialize(name, weight, price)
    @name, @weight = name, weight
    @price = if Time.now.wday == 3 and not @name.match(/egenkomponert/i)
      119
    else
      price
    end
  end

  def to_s
    "%-60s (NOK %3d)" % [@name, @price]
  end
end

class Sauce
  def self.price
    20
  end
  def self.name
    "Luguber rømmedressing"
  end
end

class Menu
  def self.add(pizza)
    @@contents ||= []
    @@contents << pizza
  end
  def self.weighted_list
    @@contents.inject([]) do |arr,elem|
      arr << [elem] * elem.weight
      arr.flatten
    end.sort_by { rand }
  end
end

class Order
  class << self
    alias_method :for, :new
  end
  def initialize(n)
    @people = n
    @hunger = 0.4
    @max    = 1200
    @total  = 0
    @items  = []
    @sauce  = 0
    sauce_prob = 0.7

    possible = Menu.weighted_list
    (@people * @hunger).to_i.times do
      pizza = possible[rand(possible.length)]
      break if @total + pizza.price > @max
      @total += pizza.price
      @items << pizza
      if rand < sauce_prob and @total + Sauce.price < @max
        @sauce += 1
        @total += Sauce.price
      end
    end
  end

  def to_s
    orderlines = @items.map {|i| i.to_s}.join("\n")
    sauceline  = "%2d x %-55s (NOK %3d)" % [@sauce, Sauce.name, Sauce.price]
    orderlines + "\n" + sauceline + "\n\n" + "TOTAL: #{@total}"
  end
end

def Pizza(name, weight, price)
  Menu.add(Pizza.new(name, weight, price))
end

Pizza("Pappas spesial", 4, 159)
Pizza("Texas",          3, 149)
Pizza("Blue Hawai",     7, 149)
Pizza("Floriad",        4, 149)
Pizza("Buffalo",        4, 149)
Pizza("Chicken",        4, 149)
Pizza("New York",       0, 149)
Pizza("Las Vegas",      6, 149)
Pizza("Vegetarianer",   0, 149)
Pizza("FILADELFIA",     4, 149)
Pizza("Hot Chicago",    7, 149)
Pizza("Hot Express",    5, 149)
Pizza("Kebab pizza spesial", 3, 169)
Pizza("Egenkomponert Pepperoni Biff Bacon Skinke løk",       9, 159)
Pizza("Egenkomponert Biff Pepperoni Bacon Skinke Tacokjøtt", 9, 159)

peeps = if ARGV.first.nil?
          print "# peeps: "
          Integer(gets)
        else
          Integer(ARGV.shift)
        end

puts Order.for(peeps)

File: pizza.erl
% Erlang version, author: Trygve Laugstøl
% erl pizza.erl
% c(pizza).
% pizza:run(8).  % Order for 8 people.
-module(pizza).
 
-export([run/0, run/1]).
 
-import(random,[seed/0,uniform/1]).
-import(lists,[duplicate/2, nth/2]).
 
-define(PRIZE_MAX, 1200).
-define(EL_CHEAPO, 119).
-define(PIZZA_PER_HEAD, 0.4).
 
-define(DEFAULT_HEADS, 5).
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% API
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
run() ->
    run(?DEFAULT_HEADS).
 
run(Heads) when is_integer(Heads) ->
    NumOfPizzas = number_of_pizzas(Heads),
    Choices = calculate_choices(menu(), []),
    Selection = select_pizzas(NumOfPizzas, Choices, []),
    print_selection(Selection).
 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Private
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
 
number_of_pizzas(Heads) ->
    C = Heads * ?PIZZA_PER_HEAD,
    T = erlang:trunc(C),
    case (C - T) of
        Neg when Neg < 0 -> T - 1;
        Pos when Pos > 0 -> T + 1;
        _ -> T
    end.
 
menu() ->
    [
        {"Pappas spesial", 4, 159},
        {"Texas", 3, 149},
        {"Blue Hawai", 7, 149},
        {"Floriad", 4, 149},
        {"Buffalo", 4, 149},
        {"Chicken", 4, 149},
        {"New York", 0, 149},
        {"Las Vegas", 6, 149},
        {"Vegetarianer", 0, 149},
        {"FILADELFIA", 4, 149},
        {"Hot Chicago", 7, 149},
        {"Hot Express", 5, 149},
        {"Kebab pizza spesial", 3, 169},
        {"Egenkomponert; Pepperoni, Biff, Bacon, Skinke, l0k", 9, 159},
        {"Egenkomponert; Biff, Pepperoni, Bacon, Skinke, Tacokj0tt", 9, 159}
    ].
 
calculate_choices([Pizza | Menu], Choices) ->
    Weight = element(2, Pizza),
    X = duplicate(Weight, Pizza),
    calculate_choices(Menu, X ++ Choices);
 
calculate_choices([], Choices) ->
    Choices.
 
select_pizzas(0, _, Selection) ->
    Selection;
 
select_pizzas(NumOfPizzas, Choices, Selection) when NumOfPizzas > 0 ->
    Selected = nth(uniform(length(Choices)), Choices),
    select_pizzas(NumOfPizzas - 1, Choices, [Selected|Selection] ).
 
print_pizzas([], _, Cost) ->
    Cost;
 
print_pizzas([Pizza | Selection], Index, TotalCost) ->
    Cost = element(3, Pizza),
    io:format("#~2..0w ~-30s ~5w~n", [Index, element(1, Pizza), Cost]),
    print_pizzas(Selection, Index + 1, TotalCost + Cost).
 
print_selection(Selection) ->
    io:format("========================================~n"),
    Cost = print_pizzas(Selection, 0, 0),
    io:format("========================================~n"),
    io:format("Count: ~2w, Cost:                   ~5w~n", [length(Selection), Cost]).
